perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java

```diff
@@ -1859,6 +1859,26 @@ public static AuthzRoles addAllSubgroupsToAuthzRoles(PerunSession sess, AuthzRol
 			groupAdminRoles.put("Group", newGroupsIds);
 			authzRoles.put(Role.GROUPADMIN, groupAdminRoles);
 		}
+		if (authzRoles.hasRole(Role.GROUPOBSERVER)) {
+			Map<String, Set<Integer>> groupObserverRoles = authzRoles.get(Role.GROUPOBSERVER);
+			Set<Integer> groupsIdsForObserver = groupObserverRoles.get("Group");
+			Set<Integer> newGroupsIdsForObserver = new HashSet<>(groupsIdsForObserver);
+			for (Integer id : groupsIdsForObserver) {
+				Group parentGroupForObserver;
+				try {
+					parentGroupForObserver = getPerunBl().getGroupsManagerBl().getGroupById(sess, id);
+				} catch (GroupNotExistsException ex) {
+					log.debug("Group with id=" + id + " not exists when initializing rights for user: " + sess.getPerunPrincipal().getUser());
+					continue;
+				}
+				List<Group> subGroups = getPerunBl().getGroupsManagerBl().getAllSubGroups(sess, parentGroupForObserver);
+				for (Group g : subGroups) {
+					newGroupsIdsForObserver.add(g.getId());
+				}
+			}
+			groupObserverRoles.put("Group", newGroupsIdsForObserver);
+			authzRoles.put(Role.GROUPOBSERVER, groupObserverRoles);
+		}
```

Comment:
I think this could be done without such a duplicate of the existing code. The only change is the role itself here.

Reply:
Changes - I did it without duplication.